### ISSUE TITLE
Inherited javax.validation constraints are ignored

### ISSUE THREAD

**Issue by jj-sn** [victools/jsonschema-generator#297]
:wave:

It is my understanding that when a getter overrides a method on an interface or a super class, all `javax.validation` constraints on the overridden method also apply to the overriding method (at least that's the behavior I observed).

However, `jsonschema-module-javax-validation` doesn't seem to pick up constraints that are inherited this way. See this failing test:

```java
    @Test
    public void schemaContainsInheritedConstraints() {
        var configBuilder = new SchemaGeneratorConfigBuilder(DRAFT_2019_09)
                .with(new JavaxValidationModule());
        var generator = new SchemaGenerator(configBuilder.build());

        var schema = generator.generateSchema(Book.class);
        var props = (ObjectNode) schema.get("properties");
        var title = (ObjectNode) props.get("title");
        assert title.get("minLength").asInt() == 1;
    }

    class Book implements Publication {
        private String title;

        @Override
        public String getTitle() {
            return title;
        }
    }

    interface Publication {
        @Size(min = 1)
        String getTitle();
    }
```

Is this a bug or am I on the wrong track?

**Comment by CarstenWickner** [comment]
Hi @jj-sn,

I finally got around to looking at this topic, just to realize it was possible all along. 😅
You merely need to declare an alternative `TypeContext` with more lenient annotation lookup configuration:
```java
        // the default AnnotationInclusion.INCLUDE_AND_INHERIT_IF_INHERITED requires @Inherited on annotations being looked up
        // using AnnotationInclusion.INCLUDE_AND_INHERIT instead considers all annotations on overridden methods as well
        SchemaGenerator generator = new SchemaGenerator(configBuilder.build(),
                TypeContextFactory.createTypeContext(AnnotationInclusion.INCLUDE_AND_INHERIT, config));
```

I've added the complete example here:
https://github.com/victools/jsonschema-generator/blob/main/jsonschema-examples/src/main/java/com/github/victools/jsonschema/examples/AnnotationInheritanceExample.java
with the output containing the desired `minLength` keyword:
https://github.com/victools/jsonschema-generator/blob/main/jsonschema-examples/src/test/resources/com/github/victools/jsonschema/examples/AnnotationInheritanceExample-result.json
```json
{
    "$schema" : "https://json-schema.org/draft/2019-09/schema",
    "type" : "object",
    "properties" : {
        "title" : {
            "type" : "string",
            "minLength" : 1
        }
    }
}
```

Seeing that the `@Inherited` annotation is not present on the javax/jakarta validation annotations, I'm actually unsure what the correct behavior should be here.

@jochenberger what do you reckon? You already took a stab at implementing the consideration of overridden methods in #299. Are the validation annotations actually inherited without the `@Inherited` marker?

**Comment by jochenberger** [comment]
@CarstenWickner, TBH, I'm not sure what you're asking. I haven't had a chance to have a look at my PR since you reviewed it, sorry.
It looks as if the validation annotations are being treated as @Inherited although they don't have that annotation. I wonder if that's part of the spec.
The `TypeContext` solution looks pretty elegant to me, but it might have (possibly unwanted) effects for other annotations, right?

**Comment by CarstenWickner** [comment]
Yes, the TypeContext setting applies globally.
One would have to extend it further to apply different settings at different times.

Doesn't sound fun but possible I suppose.
For now, one would have to weigh up what is preferred.
Maybe extending the `TypeContext` would be a cleaner solution then.

**Comment by jochenberger** [comment]
https://jakarta.ee/specifications/bean-validation/3.0/jakarta-bean-validation-spec-3.0.html#constraintdeclarationvalidationprocess-inheritance seems to suggest that validation constraints are always inherited, so maybe that's what we should to as well?

**Comment by CarstenWickner** [comment]
In #344 I'm introducing the ability to apply overrides for these settings in the `TypeContext`.
This way, the javax and jakarta validation modules can define the annotations should be inherited regardless of any `@Inherited` annotation on them, while other annotations are only considered to be inherited, when they are explicitly marked as such.

While reading through the linked documentation above, I also noticed, that the current handling of validation groups didn't consider their inheritance correctly either. That's being fixed in the same PR then.

**Comment by CarstenWickner** [comment]
Released in [v4.31.0](https://github.com/victools/jsonschema-generator/releases/tag/v4.31.0) just now. 🚀
